mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/mail/InMemoryMessageIdMapper.java

```diff
@@ -0,0 +1,148 @@
+                .collect(Guavate.toImmutableList());
+        } catch (MailboxException e) {
+            throw Throwables.propagate(e);
+        }
+
+    }
+
+    @Override
+    public List<MailboxId> findMailboxes(MessageId messageId) {
+        return find(ImmutableList.of(messageId), MessageMapper.FetchType.Metadata)
+            .stream()
+            .map(MailboxMessage::getMailboxId)
+            .collect(Guavate.toImmutableList());
+    }
+
+    @Override
+    public void save(MailboxMessage mailboxMessage) throws MailboxException {
+        Mailbox mailbox = mailboxMapper.findMailboxById(mailboxMessage.getMailboxId());
+        messageMapper.save(mailbox, mailboxMessage);
+    }
+
+    @Override
+    public void copyInMailbox(MailboxMessage mailboxMessage) throws MailboxException {
+        boolean isAlreadyInMailbox = findMailboxes(mailboxMessage.getMessageId()).contains(mailboxMessage.getMailboxId());
+        if (!isAlreadyInMailbox) {
+            save(mailboxMessage);
+        }
+    }
+
+    @Override
+    public void delete(MessageId messageId) {
+        find(ImmutableList.of(messageId), MessageMapper.FetchType.Metadata)
+            .forEach(Throwing.consumer(
+                message -> messageMapper.delete(
+                    mailboxMapper.findMailboxById(message.getMailboxId()),
+                    message)));
+    }
+
+    @Override
+    public void delete(MessageId messageId, List<MailboxId> mailboxIds) {
+        find(ImmutableList.of(messageId), MessageMapper.FetchType.Metadata)
+            .stream()
+            .filter(message -> mailboxIds.contains(message.getMailboxId()))
+            .forEach(Throwing.consumer(
+                message -> messageMapper.delete(
+                    mailboxMapper.findMailboxById(message.getMailboxId()),
+                    message)));
+    }
+
+    @Override
+    public Map<MailboxId, UpdatedFlags> setFlags(MessageId messageId, List<MailboxId> mailboxIds,
+                                                 Flags newState, FlagsUpdateMode updateMode) throws MailboxException {
+        return find(ImmutableList.of(messageId), MessageMapper.FetchType.Metadata)
+            .stream()
+            .filter(message -> mailboxIds.contains(message.getMailboxId()))
+            .map(updateMessage(newState, updateMode))
+            .collect(Guavate.entriesToMap());
+    }
+
+    private Function<MailboxMessage, Pair<MailboxId, UpdatedFlags>> updateMessage(Flags newState,
+                                                                                          FlagsUpdateMode updateMode) {
+        return Throwing.function((MailboxMessage message) -> {
+            FlagsUpdateCalculator flagsUpdateCalculator = new FlagsUpdateCalculator(newState, updateMode);
+            if (flagsUpdateCalculator.buildNewFlags(message.createFlags()).equals(message.createFlags())) {
+                return Pair.of(message.getMailboxId(),
+                    UpdatedFlags.builder()
```

Comment:
Weird indent

Reply:
I would have extracted the building flag part's to a variable

Reply:
fixed